src/Workspaces/Remote/Core/SolutionAssetProvider.cs

```diff
@@ -52,58 +52,143 @@ public async ValueTask GetAssetsAsync(PipeWriter pipeWriter, Checksum solutionCh
+                => await writer.FlushAsync(cancellationToken).ConfigureAwait(false);
+
+            public override void Write(byte[] buffer, int offset, int count)
+            {
+                Requires.NotNull(buffer, nameof(buffer));
+                Verify.NotDisposed(this);
+
+                var span = this.writer.GetSpan(count);
+                buffer.AsSpan(offset, count).CopyTo(span);
+                this.writer.Advance(count);
+            }
+
+            public override Task WriteAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+                this.Write(buffer, offset, count);
+                return Task.CompletedTask;
+            }
+
+            public override void WriteByte(byte value)
             {
-                try
-                {
-                    var stream = localPipe.Writer.AsStream(leaveOpen: false);
-                    using var writer = new ObjectWriter(stream, leaveOpen: false, cancellationToken);
-                    RemoteHostAssetSerialization.WriteData(writer, singleAsset, assetMap, serializer, scope.ReplicationContext, solutionChecksum, checksums, cancellationToken);
-                }
-                catch (Exception e) when (FatalError.ReportAndCatchUnlessCanceled(e, cancellationToken))
-                {
-                    // no-op
-                }
-            }, mustNotCancelToken);
-
-            // Complete RPC once we send the initial piece of data and start waiting for the writer to send more,
-            // so the client can start reading from the stream. Once CopyPipeDataAsync completes the pipeWriter
-            // the corresponding client-side pipeReader will complete and the data transfer will be finished.
-            var task2 = CopyPipeDataAsync();
-
-            await Task.WhenAll(task1, task2).ConfigureAwait(false);
-
-            async Task CopyPipeDataAsync()
+                Verify.NotDisposed(this);
+                var span = this.writer.GetSpan(1);
+                span[0] = value;
+                this.writer.Advance(1);
+            }
+
+#if !NETSTANDARD
+
+            public override void Write(ReadOnlySpan<byte> buffer)
+            {
+                Verify.NotDisposed(this);
+                var span = this.writer.GetSpan(buffer.Length);
+                buffer.CopyTo(span);
+                this.writer.Advance(buffer.Length);
+            }
+
+            public override ValueTask WriteAsync(ReadOnlyMemory<byte> buffer, CancellationToken cancellationToken)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+                this.Write(buffer.Span);
+                return default;
+            }
+
+#endif
+
+            protected override void Dispose(bool disposing)
+            {
+                this.IsDisposed = true;
+                base.Dispose(disposing);
```

Comment:
As this `Stream` owns the `PipeWriter` now, it seems that disposing the stream should call `PipeWriter.Complete()`. Without this, the reader side will never exit the read loop.

You may also want to define a method on this class to complete the writer with an exception so you can communicate that if needed.

Reply:
taht seems reasonable :)

Reply:
ah, so looking at this, the real owner of hte writer is the thing that calls GetAssetsAsync.  it does do what you're saying.  and a followup PR will clean that up further.  I'll add a comment to this effect.